libs/util/src/lib/generate-leaderboard.ts

```diff
@@ -0,0 +1,141 @@
+import {
+  AppFinishedPlayerValues,
+  AppLeaderboard,
+  AppLeaderboardEntry,
+  AppPlayerId,
+  AppPlayerLog,
+  AppPlayerLogId,
+  AppPlayerLogs,
+  AppPlayerStatus,
+  AppRaceId,
+  AppTimeoutPlayerValues,
+} from '@razor/models';
+import { extractId, extractIdType } from './extractIds';
+
+export const generateLeaderboard = (
+  playerLogs: AppPlayerLogs,
+  raceId: AppRaceId,
+  raceTextLength: number,
+): AppLeaderboard => {
+  const leaderboardEntries: AppLeaderboard = [];
+
+  for (const playerLogId in playerLogs) {
+    const raceIdOfPlayerLog: AppRaceId = extractId(
+      playerLogId,
+      extractIdType.playerLog,
+      extractIdType.race,
+    ) as AppRaceId;
+
+    const playerIdOfPlayerLog: AppPlayerId = extractId(
+      playerLogId,
+      extractIdType.playerLog,
+      extractIdType.player,
+    ) as AppPlayerId;
+
+    const playerLogsLength = playerLogs[playerLogId as AppPlayerLogId].length;
+    const playerLastTextLength =
+      playerLogs[playerLogId as AppPlayerLogId][playerLogsLength - 1]
+        .textLength;
+
+    let wpm, elpasedTime, distance;
+    let finishedPlayerValues: AppFinishedPlayerValues = {
+      wpm: 0,
+      elpasedTime: 0,
+    };
+    let timeoutPlayerValues: AppTimeoutPlayerValues = {
+      distance: 0,
+    };
+    //Check whether race owns the racelog
+    if (raceIdOfPlayerLog === raceId) {
+      // Check whether player has finished the race
+      if (playerLastTextLength === raceTextLength) {
+        wpm = calculateWPM(
+          raceTextLength,
+          playerLogs[playerLogId as AppPlayerLogId],
+        );
+        elpasedTime =
+          playerLogs[playerLogId as AppPlayerLogId][playerLogsLength - 1]
+            .timestamp - playerLogs[playerLogId as AppPlayerLogId][0].timestamp;
+        finishedPlayerValues = {
+          wpm,
+          elpasedTime,
+        };
+      } else {
+        distance =
+          playerLogs[playerLogId as AppPlayerLogId][playerLogsLength - 1]
+            .textLength -
+          playerLogs[playerLogId as AppPlayerLogId][0].textLength;
+        timeoutPlayerValues = { distance };
+      }
+    }
+    const leaderboardEntry: AppLeaderboardEntry = {
+      playerId: playerIdOfPlayerLog,
+      status:
+        playerLastTextLength === raceTextLength
+          ? AppPlayerStatus.Complete
+          : AppPlayerStatus.Timeout,
+      values:
+        playerLastTextLength === raceTextLength
+          ? finishedPlayerValues
+          : timeoutPlayerValues,
+    };
+    leaderboardEntries.push(leaderboardEntry);
+  }
+  return leaderboardEntries;
+};
+
+const calculateWPM = (length: number, logs: AppPlayerLog[]): number => {
```

Comment:
Can you please add documentation about this algorithm?

Reply:
Completed docs  in PR #14 under commit 4d6c57dec3d53ee4b0ea4ee596fa7368c3a0833d